shared/lsp/registration.ts

```diff
@@ -363,203 +361,106 @@ export async function register({
         fn: (connection: LSPConnection) => Promise<R>
     ) => Promise<R>
 
-    if (supportsWorkspaceFolders) {
-        const connection = await connect({
-            clientRootUri: null,
+    // Supports only one workspace root
+    // TODO this should store a refcount to avoid closing connections other consumers have a reference to
+    /** Map from client root URI to connection */
+    const connectionsByRootUri = new Map<string, Promise<LSPConnection>>()
+    withConnection = async (workspaceFolder, fn) => {
+        let connection = await connectionsByRootUri.get(workspaceFolder.href)
+        if (connection) {
+            return await fn(connection)
+        }
+        const serverRootUri = clientToServerURI(workspaceFolder)
+        connection = await connect({
+            clientRootUri: workspaceFolder,
             initParams: {
                 processId: null,
-                rootUri: null,
+                rootUri: serverRootUri.href,
                 capabilities: clientCapabilities,
-                workspaceFolders: sourcegraph.workspace.roots.map(
-                    toLSPWorkspaceFolder(clientToServerURI)
-                ),
+                workspaceFolders: null,
                 initializationOptions,
             },
-            registerProviders: true,
+            registerProviders: false,
         })
         subscriptions.add(connection)
-        withConnection = async (workspaceFolder, fn) => {
-            let tempWorkspaceFolder: lsp.WorkspaceFolder | undefined
-            // If workspace folder is not known yet, add it
-            if (
-                !sourcegraph.workspace.roots.some(
-                    root => root.uri.toString() === workspaceFolder.href
-                )
-            ) {
-                tempWorkspaceFolder = {
-                    uri: workspaceFolder.href,
-                    name: path.posix.basename(workspaceFolder.pathname),
-                }
-                connection.sendNotification(
-                    lsp.DidChangeWorkspaceFoldersNotification.type,
-                    {
-                        event: {
-                            added: [tempWorkspaceFolder],
-                            removed: [],
-                        },
-                    }
-                )
-            }
-            try {
-                return await fn(connection)
-            } finally {
-                // If workspace folder was added, remove it
-                if (tempWorkspaceFolder) {
-                    connection.sendNotification(
-                        lsp.DidChangeWorkspaceFoldersNotification.type,
-                        {
-                            event: {
-                                added: [],
-                                removed: [tempWorkspaceFolder],
-                            },
-                        }
+        try {
+            return await fn(connection)
+        } finally {
+            connectionsByRootUri.delete(workspaceFolder.href)
```

Comment:
I think I'm missing something. Reaching 373 appears to mean `connectionsByRootUri` did not contain `workspaceFolder.href`, or else we would have exited at 371. Should we have added it before 386?

Reply:
This can throw async. We put the _promise_ in the map and then if the promise fails during init we'll remove ourselves.

Reply:
Right—and during `addRoots` it works that way. But here we aren't adding any mass to the map, and we've already ruled out that it was present before.

What I'm not following is that in the case where we _did_ find the promise in the map, we didn't remove it when we're done; but here (where it's not) we do.  I realize the promise may have failed between creation and now, but the registration in the map happens sooner as far as I can see.

Reply:
Oops, sorry. I was thinking this was another similar line in the file. You're right that this is unnecessary.